### PR TITLE
Add basic_dao rust example

### PR DESCRIPTION
**Pull request by rstout**

### CODE REVIEW

**Comment by crusso** [comment]
Looking good but can you add a github workflow to verify it builds and has a basic smoke test, or a more significant one like #140 . That way we can avoid bit-rot and test new release of dfx.

BTW does this actually implement the same Candid interface as #140. Probably easier for me to just ask than manually verify ;->.

If #140 and this are part of the same example, maybe it would make sense to merge the PRs and share tests etc.

File: rust/basic_dao/README.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
+# Basic DAO
+
+This sample project demonstrates a basic DAO (Decentralized Autonomous Organization) that can be deployed to the
+[Internet Computer](https://github.com/dfinity/ic) 
+
```
Comment (crusso) [inline rust/basic_dao/README.md:5]: I'd explain briefly what a DAO is, here or in the next section. Or point to wikipedia if there is a good description.

File: rust/basic_dao/README.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+* You have stopped any Internet Computer or other network process that would
+   create a port conflict on 8000.
+
+## Demo
+
+1. Build the `basic_dao` canister.
+
+   ```text
+   $ make clean; make
```
Comment (crusso) [inline rust/basic_dao/README.md:41]: Doesn't dfx build the rust project for you?

File: rust/basic_dao/dfx.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+      "build": [
+        "cargo build --target wasm32-unknown-unknown --package  basic_dao --release",
+        "ic-cdk-optimizer target/wasm32-unknown-unknown/release/basic_dao.wasm -o target/wasm32-unknown-unknown/release/basic_dao_opt.wasm"
+      ],
+      "candid": "src/basic_dao/src/basic_dao.did",
+      "wasm": "target/wasm32-unknown-unknown/release/basic_dao_opt.wasm",
+      "type": "custom"
+    }
+  }
+}
```
Comment (crusso) [inline rust/basic_dao/dfx.json:13]:
```suggestion
}

```

File: rust/basic_dao/src/basic_dao/src/basic_dao.did [hunk 4]
```diff
@@ -0,0 +1,126 @@
+type BasicDaoStableStorage = record {
```
Comment (crusso) [inline rust/basic_dao/src/basic_dao/src/basic_dao.did:1]: I would link to this file from the README so users can easily see the interface.

File: rust/basic_dao/src/basic_dao/src/env.rs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    fn caller(&self) -> Principal {
+        self.caller
+    }
+
+    fn canister_id(&self) -> Principal {
+        self.canister_id
+    }
+}
```
Comment (crusso) [inline rust/basic_dao/src/basic_dao/src/env.rs:64]:
```suggestion
}

```

**Comment by rstout** [comment]
> Looking good but can you add a github workflow to verify it builds and has a basic smoke test, or a more significant one like #140 . That way we can avoid bit-rot and test new release of dfx.

I don't have any more cycles to commit to basic_dao, any additional work will require a ticket, prioritization, etc.

> BTW does this actually implement the same Candid interface as #140. Probably easier for me to just ask than manually verify ;->.

Yes. However, they are slightly out of sync currently (changes were made to the Rust impl/iface after the Motoko one was finished).

> If #140 and this are part of the same example, maybe it would make sense to merge the PRs and share tests etc.

This would be a lot easier if there was a single dir, e.g. `examples/basic_dao/{motoko, rust, ui, tests, etc}`

> I noticed that the Motoko version supports upgrades - does this one?

No

File: rust/basic_dao/README.md (hunk 2 above)
Comment (rstout) [reply rust/basic_dao/README.md:42]: `dfx build` either selects or expects canister IDs to exist. `make` and then `dfx deploy` is the best build/deploy path I found.

**Comment by rstout** [comment]
@crusso Could we get this merged this week?

**Comment by crusso** [comment]
I don't think that's up to me. Someone needs to address the issue in both (ie. make sure the rust and motoko version expose the same interface/do the same thing) and add CI so this stuff can be tested regularly without bit-rotting. Has that been done?

**Comment by crusso** [comment]
@rstout @chenyan-dfinity can you get together, merge #142 and #140 into a single PR with a consistent interface and working CI for both?

File: rust/basic_dao/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+[workspace]
+members = [
+    "src/basic_dao"
+]
```
Comment (crusso) [inline rust/basic_dao/Cargo.toml:4]:
```suggestion
]

```

**Review by crusso — APPROVED** [review]
Good stuff. Clear code and Readme (not a Rustian myself so very superficial review).

There's no reference to the ledger at all so I guess this is just toy tokens. Was that the remit for the project?

It would be great if you could eventually turn the Readme script into a CI test (see the other .yml files in examples/.github/workflows/ for guidance). Even minimal ci for now to just check the project builds would be useful,

Approving in the interest of time so you can prepare for the video.

It would be good if #140 could match this API when submitted (even if that means not using Motoko Result types with lowercase `err` and `ok` tags.)

File: rust/basic_dao/README.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
+# Basic DAO
+
+This sample project demonstrates a basic DAO 
+([Decentralized Autonomous Organization](https://en.wikipedia.org/wiki/Decentralized_autonomous_organization)) 
+that can be deployed to the [Internet Computer](https://github.com/dfinity/ic)
```
Comment (crusso) [inline rust/basic_dao/README.md:5]:
```suggestion
that can be deployed to the [Internet Computer](https://github.com/dfinity/ic).
```

**Pull request merged**